### DISCUSSION TITLE
What if dumping was more human readable?

### DISCUSSION THREAD

**Discussion by sergeniously** [nlohmann/json#4117] [Ideas]
Currently, we have to make guess about what arguments of dump function mean. They are not obvious at first glance:
```c++
nlohmann::json json = {1, 2, 3};
json.dump(1, ' ', true, handler);
```
Don't you think that specifying a boolean argument as an option in a function is a bad practice?
Also, we have to provide all the arguments, if we only need to set an error handler.

I suggest a perfect solution to solve this problem.
What if there was a special class to configure a dump options?
For example:
```c++

struct nlohmann::json::dumping {
	int indent_size = -1;
	char indent_char = ' ';
	bool ensure_ascii = false;
	error_handler_t error_handler;

	format& with_indent_size(int) { ... }
	format& with_indent_char(char) { ... }
	format& with_ensure_ascii(bool = true) { ... }
	format& with_error_handler(error_handler_t) { ... }
}

// ....

using nlohmann::json;

json json = {1, 2, 3};
json.dump(json::dumping().with_indent_size(1).with_error_handler(handler));
```
This way we could individually set an error handler without specifying all other options.
And it's perfectly human readable.
Think about it.

**Comment by nlohmann** [comment]
The `dump` interface grew over the years, so it's unfortunately not in the best shape. I understand your approach (I think we already had some issues discussion similar ideas), but I am not sure if it is compatible in the long run.
